sw/device/silicon_creator/lib/cert/dice_cwt.c

```diff
@@ -310,10 +322,140 @@ rom_error_t dice_cdi_1_cert_build(hmac_digest_t *owner_measurement,
+  *y_coord_p = NULL;
+
+  // Locate public key coords bstrs.
+  int64_t key = 0;
+  for (size_t i = 0; i < num_pairs; ++i) {
+    CborReadInt(&cbor_obj, &key);
+    if (key == kDiceCwtCoseKeyXCoordLabel) {
+      CborReadBstr(&cbor_obj, &x_coord_size, (const uint8_t **)x_coord_p);
+      continue;
+    } else if (key == kDiceCwtCoseKeyYCoordLabel) {
+      CborReadBstr(&cbor_obj, &y_coord_size, (const uint8_t **)y_coord_p);
+      continue;
+    }
+    CborReadSkip(&cbor_obj);
+  }
+
+  // Confirm we found the key.
+  if (x_coord_p == NULL || y_coord_p == NULL) {
+    return kErrorDiceCwtKeyCoordsNotFound;
+  }
+  // Confirm the key is the correct size.
+  if (x_coord_size != kEcdsaP256PublicKeyCoordBytes ||
+      y_coord_size != kEcdsaP256PublicKeyCoordBytes) {
+    return kErrorDiceCwtCoseKeyBadSize;
+  }
+
+  return kErrorOk;
+}
+
+static rom_error_t extract_pubkey_from_cose_sign1_payload(
+    const uint8_t *payload, size_t payload_size, const uint32_t **x_coord_p,
+    const uint32_t **y_coord_p) {
+  // Initialize CBOR object.
+  struct CborIn cbor_obj;
+  CborInInit(payload, payload_size, &cbor_obj);
+
+  // Find the COSE_Key object in the COSE_Sign1 payload.
+  size_t num_payload_pairs = 0;
+  CborReadMap(&cbor_obj, &num_payload_pairs);
+  if (num_payload_pairs == 0) {
+    return kErrorDiceCwtCoseKeyNotFound;
+  }
+
+  int64_t key = 0;
+  size_t cose_key_size = 0;
+  const uint8_t *cose_key_p = NULL;
+  for (size_t i = 0; i < num_payload_pairs; ++i) {
+    CborReadInt(&cbor_obj, &key);
+    if (key == kDiceCwtSubjectPublicKeyLabel) {
+      CborReadBstr(&cbor_obj, &cose_key_size, &cose_key_p);
+      break;
+    }
+    CborReadSkip(&cbor_obj);
+  }
+
+  // Check the COSE_Key object was found in the COSE_Sign1 payload.
+  if (key != kDiceCwtSubjectPublicKeyLabel) {
+    return kErrorDiceCwtCoseKeyNotFound;
+  }
+
+  // Extract the pubkey from the COSE_Key payload.
+  RETURN_IF_ERROR(extract_pubkey_from_cose_key(cose_key_p, cose_key_size,
+                                               x_coord_p, y_coord_p));
+
+  return kErrorOk;
+}
+
 rom_error_t dice_cert_check_valid(const perso_tlv_cert_obj_t *cert_obj,
                                   const hmac_digest_t *pubkey_id,
                                   const ecdsa_p256_public_key_t *pubkey,
                                   hardened_bool_t *cert_valid_output) {
-  // TODO(lowRISC/opentitan:#24281): implement body
+  const uint32_t *x_coord_p = NULL;
+  const uint32_t *y_coord_p = NULL;
+
+  struct CborIn cbor_obj;
+  CborInInit(cert_obj->cert_body_p, cert_obj->cert_body_size, &cbor_obj);
+
+  // Check if CBOR object is array or map. UDS is a COSE_Key which is a map.
+  // CDI_* are COSE_Sign1 objects, which are arrays.
+  size_t num_array_items = 0;
+  CborReadArray(&cbor_obj, &num_array_items);
+  size_t num_map_items = 0;
+  CborReadMap(&cbor_obj, &num_map_items);
+
+  // Extract the public key from the CBOR certificate object.
+  if (num_array_items > 0 && num_map_items == 0) {
```

Comment:
How about testing the `cert_obj->name == UDS` instead of this num check?

Reply:
I thought about that, however, in the future, we aim to support other DICE CWT profiles (see RFC: https://docs.google.com/document/d/1zBudM3YyasCoYhchRDLJUKKDEtlCQ4Yc5dx9HEE45HU/edit?usp=sharing), in which case the UDS might be a COSE_Sign1 object, not just. COSE_Key. so this is more flexible towards that.